rust/log-service/src/lib.rs

```diff
@@ -967,12 +792,228 @@ impl LogServer {
+        self.set_backpressure(&backpressure);
+        let mut need_to_compact = self.need_to_compact.lock();
+        std::mem::swap(&mut *need_to_compact, &mut rollups);
+        Ok(())
+    }
+
+    /// Read the entirety of a prefix of the dirty log.
+    #[tracing::instrument(skip(self), err(Display))]
+    #[allow(clippy::type_complexity)]
+    async fn read_dirty_log(
+        &self,
+    ) -> Result<(Option<Witness>, Cursor, Vec<(LogPosition, DirtyMarker)>), Error> {
+        let Some(reader) = self.dirty_log.reader(LogReaderOptions::default()) else {
+            return Err(Error::CouldNotGetDirtyLogReader);
+        };
+        let Some(cursors) = self.dirty_log.cursors(CursorStoreOptions::default()) else {
+            return Err(Error::CouldNotGetDirtyLogCursors);
+        };
+        let witness = cursors.load(&STABLE_PREFIX).await?;
+        let default = Cursor::default();
+        let cursor = witness
+            .as_ref()
+            .map(|w| w.cursor())
+            .unwrap_or(&default)
+            .clone();
+        tracing::info!("cursoring from {cursor:?}");
+        let dirty_fragments = reader
+            .scan(
+                cursor.position,
+                Limits {
+                    max_files: Some(10_000),
+                    max_bytes: Some(1_000_000_000),
+                },
+            )
+            .await?;
+        if dirty_fragments.is_empty() {
+            return Ok((witness, cursor, vec![]));
+        }
+        if dirty_fragments.len() >= 1_000 {
+            tracing::error!("Too many dirty fragments: {}", dirty_fragments.len());
+        }
+        let dirty_futures = dirty_fragments
+            .iter()
+            .map(|fragment| reader.read_parquet(fragment))
+            .collect::<Vec<_>>();
+        let dirty_raw = futures::future::try_join_all(dirty_futures).await?;
+        let mut dirty_markers = vec![];
+        for (_, records, _) in dirty_raw {
+            let records = records
+                .into_iter()
+                .flat_map(|x| match serde_json::from_slice::<DirtyMarker>(&x.1) {
+                    Ok(marker) => Some((x.0, marker)),
+                    Err(err) => {
+                        tracing::error!(
+                            "could not read marker for {}: {err}",
+                            String::from_utf8_lossy(&x.1)
+                        );
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+            dirty_markers.extend(records);
+        }
+        Ok((witness, cursor, dirty_markers))
+    }
+
+    /// Enrich a rolled up dirty log by reading cursors and manifests to determine what still needs
+    /// to be compacted.  Entries will be removed if they correspond to a compacted log range.
+    /// Entries will remain if there is data to be collected.
+    #[tracing::instrument(skip(self, rollups), err(Display))]
+    async fn enrich_dirty_log(
```

Comment:
it may make sense in a follow up PR to do the io here as low-pri in NAC. worried about 

1. serial io being slow under a high cardinality of collections
2. use NAC priority to get around blocking foreground